File: typescript/src/validate.ts
// MIT License
// Copyright (c) 2026 sparetimecoders

import type { Endpoint, Topology, Transport } from "./types.js";

/**
 * Validate checks a single service's topology for internal consistency.
 * Returns null if valid, or an error message string if invalid.
 */
export function validate(t: Topology): string | null {
  if (!t.serviceName) {
    return "service name must not be empty";
  }

  const errs: string[] = [];

  for (let i = 0; i < (t.endpoints ?? []).length; i++) {
    const ep = t.endpoints[i];

    if (!ep.exchangeName) {
      errs.push(`endpoint[${i}]: exchange name must not be empty`);
    }

    if (ep.direction === "consume" && !ep.queueName && !ep.ephemeral) {
      errs.push(
        `endpoint[${i}]: consume endpoint must have a queue name`,
      );
    }

    if (!ep.routingKey && ep.exchangeKind !== "headers") {
      errs.push(
        `endpoint[${i}]: routing key must not be empty for ${ep.exchangeKind} exchange`,
      );
    }

    if (ep.routingKey && ep.routingKey.includes(">")) {
      errs.push(
        `endpoint[${i}]: routing key must not contain '>' (use '#' for multi-level wildcard)`,
      );
    }

    const transportErr = validateExchangeNameForTransport(
      t.transport,
      ep,
      i,
    );
    if (transportErr) {
      errs.push(transportErr);
    }
  }

  return errs.length > 0 ? errs.join("\n") : null;
}

function validateExchangeNameForTransport(
  transport: Transport | undefined,
  ep: Endpoint,
  index: number,
): string | null {
  switch (transport) {
    case "amqp":
      return validateAMQPExchangeName(ep, index);
    case "nats":
      return validateNATSExchangeName(ep, index);
    default:
      return null;
  }
}

function validateAMQPExchangeName(
  ep: Endpoint,
  index: number,
): string | null {
  const name = ep.exchangeName;
  switch (ep.exchangeKind) {
    case "topic":
      if (!name.endsWith(".topic.exchange")) {
        return `endpoint[${index}]: topic exchange "${name}" must end with .topic.exchange`;
      }
      break;
    case "direct":
      if (!name.endsWith(".direct.exchange.request")) {
        return `endpoint[${index}]: direct exchange "${name}" must end with .direct.exchange.request`;
      }
      break;
    case "headers":
      if (!name.endsWith(".headers.exchange.response")) {
        return `endpoint[${index}]: headers exchange "${name}" must end with .headers.exchange.response`;
      }
      break;
  }
  return null;
}

function validateNATSExchangeName(
  ep: Endpoint,
  index: number,
): string | null {
  const name = ep.exchangeName;
  const amqpSuffixes = [
    ".topic.exchange",
    ".direct.exchange.request",
    ".headers.exchange.response",
  ];
  for (const suffix of amqpSuffixes) {
    if (name.endsWith(suffix)) {
      return `endpoint[${index}]: NATS exchange "${name}" must not use AMQP suffix "${suffix}"`;
    }
  }
  if (/[\s]/.test(name)) {
    return `endpoint[${index}]: NATS exchange "${name}" must not contain whitespace`;
  }
  return null;
}

/**
 * ValidateTopologies checks cross-service consistency across multiple topologies.
 * Returns null if valid, or an error message string if invalid.
 */
export function validateTopologies(topologies: Topology[]): string | null {
  const errs: string[] = [];

  for (const t of topologies) {
    const err = validate(t);
    if (err) {
      errs.push(`service "${t.serviceName}": ${err}`);
    }
  }

  // Group topologies by transport for cross-validation.
  const groups = new Map<string, Topology[]>();
  for (const t of topologies) {
    const key = t.transport ?? "";
    const group = groups.get(key) ?? [];
    group.push(t);
    groups.set(key, group);
  }

  for (const group of groups.values()) {
    const err = crossValidateGroup(group);
    if (err) {
      errs.push(err);
    }
  }

  return errs.length > 0 ? errs.join("\n") : null;
}

function crossValidateGroup(topologies: Topology[]): string | null {
  const errs: string[] = [];

  // Build a set of all published routing keys per exchange.
  const published = new Map<string, string>(); // "exchange|routingKey" -> service name
  for (const t of topologies) {
    for (const ep of t.endpoints ?? []) {
      if (ep.direction === "publish" && ep.routingKey) {
        published.set(
          `${ep.exchangeName}|${ep.routingKey}`,
          t.serviceName,
        );
      }
    }
  }

  // Check that each consumer has a matching publisher.
  for (const t of topologies) {
    for (const ep of t.endpoints ?? []) {
      if (ep.direction !== "consume") continue;
      if (!ep.routingKey) continue;
      // Skip wildcard consumers.
      if (/[#*>]/.test(ep.routingKey)) continue;

      const key = `${ep.exchangeName}|${ep.routingKey}`;
      if (!published.has(key)) {
        errs.push(
          `service "${t.serviceName}" consumes "${ep.routingKey}" on exchange "${ep.exchangeName}" but no service publishes it`,
        );
      }
    }
  }

  return errs.length > 0 ? errs.join("\n") : null;
}

File: validate.go
// MIT License
//
// Copyright (c) 2026 sparetimecoders
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package messaging

import (
	"errors"
	"fmt"
	"strings"
)

// Validate checks a single service's topology for internal consistency.
// It verifies structural rules (common to all transports) and transport-specific
// naming conventions when Transport is set.
func Validate(t Topology) error {
	if t.ServiceName == "" {
		return errors.New("service name must not be empty")
	}

	var errs []error
	for i, ep := range t.Endpoints {
		if ep.ExchangeName == "" {
			errs = append(errs, fmt.Errorf("endpoint[%d]: exchange name must not be empty", i))
		}
		if ep.Direction == DirectionConsume && ep.QueueName == "" && !ep.Ephemeral {
			errs = append(errs, fmt.Errorf("endpoint[%d]: consume endpoint must have a queue name", i))
		}
		if ep.RoutingKey == "" && ep.ExchangeKind != ExchangeHeaders {
			errs = append(errs, fmt.Errorf("endpoint[%d]: routing key must not be empty for %s exchange", i, ep.ExchangeKind))
		}
		if strings.Contains(ep.RoutingKey, ">") {
			errs = append(errs, fmt.Errorf("endpoint[%d]: routing key must not contain '>' (use '#' for multi-level wildcard)", i))
		}
		if err := validateExchangeNameForTransport(t.Transport, ep); err != nil {
			errs = append(errs, fmt.Errorf("endpoint[%d]: %w", i, err))
		}
	}
	return errors.Join(errs...)
}

func validateExchangeNameForTransport(transport Transport, ep Endpoint) error {
	switch transport {
	case TransportAMQP:
		return validateAMQPExchangeName(ep)
	case TransportNATS:
		return validateNATSExchangeName(ep)
	default:
		// Unknown or empty transport: skip naming checks.
		return nil
	}
}

func validateAMQPExchangeName(ep Endpoint) error {
	name := ep.ExchangeName
	switch ep.ExchangeKind {
	case ExchangeTopic:
		if !strings.HasSuffix(name, ".topic.exchange") {
			return fmt.Errorf("topic exchange %q must end with .topic.exchange", name)
		}
	case ExchangeDirect:
		if !strings.HasSuffix(name, ".direct.exchange.request") {
			return fmt.Errorf("direct exchange %q must end with .direct.exchange.request", name)
		}
	case ExchangeHeaders:
		if !strings.HasSuffix(name, ".headers.exchange.response") {
			return fmt.Errorf("headers exchange %q must end with .headers.exchange.response", name)
		}
	}
	return nil
}

func validateNATSExchangeName(ep Endpoint) error {
	name := ep.ExchangeName
	for _, suffix := range []string{".topic.exchange", ".direct.exchange.request", ".headers.exchange.response"} {
		if strings.HasSuffix(name, suffix) {
			return fmt.Errorf("NATS exchange %q must not use AMQP suffix %q", name, suffix)
		}
	}
	if strings.ContainsAny(name, " \t\n\r") {
		return fmt.Errorf("NATS exchange %q must not contain whitespace", name)
	}
	return nil
}

// ValidateTopologies checks cross-service consistency across multiple topologies.
// It verifies that consumer routing keys have matching publishers.
// Topologies are grouped by Transport before cross-validation; different
// transports cannot communicate with each other.
func ValidateTopologies(topologies []Topology) error {
	var errs []error

	for _, t := range topologies {
		if err := Validate(t); err != nil {
			errs = append(errs, fmt.Errorf("service %q: %w", t.ServiceName, err))
		}
	}

	// Group topologies by transport for cross-validation.
	groups := map[Transport][]Topology{}
	for _, t := range topologies {
		groups[t.Transport] = append(groups[t.Transport], t)
	}

	for _, group := range groups {
		if err := crossValidateGroup(group); err != nil {
			errs = append(errs, err)
		}
	}

	return errors.Join(errs...)
}

func crossValidateGroup(topologies []Topology) error {
	var errs []error

	// Build a set of all published routing keys per exchange.
	type exchangeKey struct {
		exchange   string
		routingKey string
	}
	published := make(map[exchangeKey]string) // value = publishing service name
	for _, t := range topologies {
		for _, ep := range t.Endpoints {
			if ep.Direction == DirectionPublish && ep.RoutingKey != "" {
				published[exchangeKey{ep.ExchangeName, ep.RoutingKey}] = t.ServiceName
			}
		}
	}

	// Check that each consumer has a matching publisher (exact match only;
	// wildcard consumers are expected to match a superset of publishers).
	for _, t := range topologies {
		for _, ep := range t.Endpoints {
			if ep.Direction != DirectionConsume {
				continue
			}
			if ep.RoutingKey == "" {
				continue
			}
			// Skip wildcard consumers — they intentionally match multiple keys.
			if strings.ContainsAny(ep.RoutingKey, "#*>") {
				continue
			}
			ek := exchangeKey{ep.ExchangeName, ep.RoutingKey}
			if _, ok := published[ek]; !ok {
				errs = append(errs, fmt.Errorf(
					"service %q consumes %q on exchange %q but no service publishes it",
					t.ServiceName, ep.RoutingKey, ep.ExchangeName))
			}
		}
	}

	return errors.Join(errs...)
}
